### ISSUE TITLE
Mixed tabs in C?

### ISSUE THREAD

**Issue by johalun** [helix-editor/helix#8197]
Hi All!

Is there a way to configure indent setting for mixed tabs? That i,s basic indent is 4 spaces, but whenever the indent is a multiple of 8, a tab is used instead? We have a big project that uses this style and it would be awesome if I could use Helix for it.

If not supported by Helix yet, is there something I can add in the grammar file to make this work?  

Thank you!!!

**Comment by Omnikar** [comment]
If I may ask, what's the reasoning behind this style choice? It seems odd, especially considering that some might have their tab render width to be set to something other than 8.

**Comment by johalun** [comment]
Not our choice, it's a horrible legacy we have to work with..

**Comment by johalun** [comment]
Vim and Emacs can handle it. In Vim it's `setl softtabstop=4 shiftwidth=4 noexpandtab`.

**Comment by johalun** [comment]
Is the formatter used at all when editing a file or only when formatting using for example `:fmt`? If the formatter is involved I could maybe configure a `.clang-format` to handle this.

**Comment by pascalkuthe** [comment]
I was confused here at first because you didn't post the full vim config/the explanation was a bit unclear.

What you have is that the visual width of a tab char is set to 8 (`tabstop=8`) which is the default vim setting. Using `softabwidth=4` and `shifttab=4` you configure vim to insert 4 spaces when pressing the tab (or to delete 4 spaces with backspace). While helix does allow you to configure tab width and the indent of a file separately (so `shifttab=4` and `tabstop=8` are possible). We have no equivalent to `softtabwidth`. I personally think its extremely niche (as evident that your legacy codebase is the very first issue about this).

Helix can run an autoformatter on save (you just need to turn it on, clangd can do the autoformatting) so that can easily cover the "turn 8 spaces into a tab" usecase. What that doens't cover is the usecase of turning the tab back into spaces. 

The most ergonomic way to deal with this is likely to just work entirely with spaces and to run clangfmt in a precommit hook. I am not sure if we want to support this in core since it does seem really niche and add quite a bit of extra completiity to tab.

**Comment by johalun** [comment]
How would I do `shifttab=4` and `tabstop=8`  in Helix? I haven't been able to find any information on indent options other than `tab-width` and `unit`. 

Yeah I understand it's a niche style by now. Re-styling the code is not an option. For now I went back to Spacemacs..

**Comment by pascalkuthe** [comment]
tabstop=8 and shifttab=4 would be equivaent to setting `unit` to 4 spaces and tab-width to 8

**Comment by johalun** [comment]
Ah I see. Yeah I tried that and while it displays the code correctly, editing and inserting new code doesn't get correctly indented.

**Comment by johalun** [comment]
My hopes are that we can do a minimal whitespace change and convert tabs to spaces to save our sanity, and enable me to go back to my favorite editor :)

**Comment by pascalkuthe** [comment]
Yeah I thought about it a bit more and this just seems way to nieche to be supported in core. 

Another possible workaround for you could be to have gjt books that convert tabs to spaces in checked and then convert back on check in
